### ISSUE TITLE
Optional Fields

### ISSUE THREAD

**Issue by karhu** [USCiLab/cereal#30]
I was wondering if support for optional serialization fields is considered/planned.

In my case I would use this feature when loading/writing settings files from/to json (or xml), where certain entries don't have to be present. But also in the scenario of binary serialization this could lead to memory savings.

Based on the boost_compat_new branch I was able to implement this for my own maybe<T> type and the json archives. For this I had to add another variant of the templated save / load functions of the json-archive and type specific prologue/epilogue functions.

Is there any interest in officially supporting something similar?

**Comment by aseering** [comment]
The discussion above seems to imply a straightforward implementation -- implement a load() method that catches and appropriately handles the exception that's thrown in case the field is missing.  Is that what y'all're thinking of doing?  Or did you have deeper changes in mind?

**Comment by heinermann** [comment]
The problem with catching exceptions is that the performance will be significantly impacted. And when I write significantly, I'm not exaggerating.

**Comment by aseering** [comment]
Indeed...  Is there a cheaper way to intercept this failure currently in Cereal?
On Mar 31, 2016 10:03 PM, Adam Heinermann <notifications@github.com> wrote:The problem with catching exceptions is that the performance will be significantly impacted. And when I write significantly, I'm not exaggerating.

**Comment by Type1J** [comment]
The experience from Protocol Buffer, Cap'n Proto, etc. explicitly states that treating missing fields as an error is a bad thing.

However, the other side of that argument is that state that is expected to be modified on load will not be, and instead would retain its previous value. However, if the object always resets its state on load to some default state, then it wouldn't be a problem.

I'd say that it's easier to work around problems caused by having all fields optional than working around all fields being required. Because it's easier, the exception throw on not finding a field to which a deserialized value can bind should be removed, and Cereal should increment its version number.

@gramic I've been looking for a BSON Cereal archive. Have you released the BSON Cereal archive that you talked about above? If so, where could I get it?

**Comment by gramic** [comment]
@Type1J I didn't open source BSON Cereal archive, mainly because I use very old versions of both BSON and Cereal libraries. Well, also some of the code is not efficient as much as it should be to show off.

Recently, MongoDB did create a very modern c++11 libraries, which will be perfect for use in a new modern version of BSON Cereal archive that I intent to use and then open source.

About the missing optional field in Cereal, my current use is to **require** every field to be present. That forces me to update all document fields in MongoDB before using the newly added field via the BSON Cereal archive. A bit inconvenient, but it works.

**Comment by aseering** [comment]
@Type1J  for what it's worth, I've worked with a library where missing fields are not an error, and that can be a bad thing too :-)  I think both extremes have significant problems.

I wouldn't mind if, for performance reasons, there were a return code rather than an exception.  Or maybe a callback function that could compute a value or do something else.

**Comment by Type1J** [comment]
@gramic I'll be interested in that archive when you get around to it.

@aseering I've always been bitten by required fields in evolving protocols. Optional fields can be handled manually, but rejecting something because it had a field that might actually be insignificant is probably bad, and the class should probably decide if it really is bad or not. Throwing an exception because something wasn't found is probably acceptable, but even then, it should be optional if the class says "Call this function with each field that wasn't found in the deserialized object" by registering a callback.

Since each use of Cereal will likely have a different set of opinions, I'd say that each object should probably register its own personal opinion of what will be, and what will not be accepted as a serialized object.

**Comment by Enhex** [comment]
The exception solution is brittle, and doesn't work in some cases. [example of a bad case](https://gist.github.com/Enhex/293382d1504b6a1febcdb45569aa7a15)

Another solution that is exception-free and can handle the example I gave is to loop over the existing node names:
```C++
load(...)
{
	while (true)
	{
		const auto namePtr = archive.getNodeName();

		if (!namePtr)
			break;

		const auto name = std::string(namePtr);

		if(name == /*...*/)
			archive(cereal::make_nvp(name, /*...*/));
		else if(name == /*...*/)
			archive(cereal::make_nvp(name, /*...*/));
		//...
	}
}
```

**Comment by rcdailey** [comment]
Wow I'm surprised this issue has been unresolved for 3 years. This is a major setback for me. I started using cereal for HTTP response body parsing and some response properties (in JSON) that I get back may not be present.

In my humble opinion, if the field is missing all cereal has to do is quietly return without mutating the variable passed into make_nvp(). In this case, whatever value that member had when the object was constructed will be what is used by consumers of the serialized object.

Note this is for *loading* serialized data (JSON input archive).

**Comment by erichkeane** [comment]
I did an optional value at one point like this (where "v" is the value):

    try
        {
            ar(v);
        }
        catch(cereal::Exception&)
        {
            ar.setNextName(nullptr);
            // Loading a key that doesn't exist results in an exception
            // Since "Not Found" is a valid condition for us, we swallow
            // this exception and the archive will not load anything
    }

**Comment by joshhaines** [comment]
@erichkeane This is exactly what I was looking for in my case. We have a bunch of optional values we are try/catching in a vector of structs. I noticed if the last object failed deserializing in the load block that it would just not even bother deserializing the rest of the structs in the vector. Clearing nextName with ar.setNextName(nullptr) was exactly what I was looking for. Thanks!

**Comment by Enhex** [comment]
I wrote a header that provides an elegant solution for optional NVPs:
https://github.com/Enhex/Cereal-Optional-NVP

**Comment by LowCostCustoms** [comment]
Here I wrote some methods to make work with optional values easier.
https://github.com/LowCostCustoms/cereal-optional

**Comment by Dysl3xik** [comment]
@AzothAmmo If you guy are willing to add a simple function to the XML && JSON archives this problem becomes really simple.

As a public function in XMLInputArchive (mostly repeat code from Start Node but WITHOUT throwing):
`
		//! Check if node with given name exists
		inline bool hasName(const char * searchName)
		{
			auto next = itsNodes.top().child; 
			if (searchName && (next == nullptr || std::strcmp(next->name(), searchName) != 0))
			{
				next = itsNodes.top().search(searchName);
			}
			return next != nullptr;
		}
`
As a public function in JSONInputArchive (most repeat code from search but WITHOUT throwing!):
'
inline bool hasName(const char* name)
		{
			// The name an NVP provided with setNextName()
			if (name)
			{
				// The actual name of the current node
				auto const actualName = itsIteratorStack.back().name();

				// Do a search if we don't see a name coming up, or if the names don't match
				if (!actualName || std::strcmp(name, actualName) != 0)
					return itsIteratorStack.back().hasName(name); //YOU NEED TO COPY SEARCH to hasName and remove the throw bits
				return true;
			}
		}
'

Finally the optional NVP similar to @Enhex but now this works out of order
Also I used C++17 here but you can convert it to template meta code without much hassle:
'
namespace cereal
{
	/// <summary>
	/// Save/Load an NVP if its name is located inside the text archive
	/// Returns TRUE if the value is loaded or saved
	/// </summary>
	template <typename Archive, typename T>
	bool make_optional_nvp(Archive& ar, const char* name, T&& value)
	{
		constexpr bool isXMLArchive = traits::is_text_archive<Archive>::value;
		constexpr bool isInputArchive = std::is_base_of_v<InputArchive<Archive>, Archive>;

		//Do check for node here, if not available then bail!
		if constexpr  (isXMLArchive && isInputArchive)
		{
			if (!ar.hasName(name))
				return false;
		}

		ar(make_nvp(name, std::forward<T>(value)));

		return true;
	}

}

#define CEREAL_OPTIONAL_NVP(ar, T) ::cereal::make_optional_nvp(ar, #T, T)
'

**Comment by Dysl3xik** [comment]
Also if you are interested in seeing this more fully fleshed out I can fork and share - but this snippet covers most of it really.

**Comment by avibahra** [comment]
I need this functionality. But are we going to see this in any release  soon ?
Also is there any c++11 solution for this.

**Comment by tankorsmash** [comment]
I'd love to see an implementation of this @Dysl3xik, having non-required fields in cereal would be great.
